### ISSUE TITLE
Encrypt mnemonic payload using PKE

### ISSUE THREAD

**Issue by im-adithya** [getAlby/nostr-wallet-connect-next#142]
See https://github.com/getAlby/nostr-wallet-connect-next/pull/84#discussion_r1524231931

**Comment by rolznz** [comment]
When the user goes through the onboarding, a mnemonic is generated locally. The user's mnemonic and unlock password are sent through an http request to the go backend to setup the server.

To be extra safe we would like to provide another layer of encryption for all HTTP requests and responses (this is not needed in Wails mode since no external requests are made). We can do this by adding a middleware to the [http service](https://github.com/getAlby/nostr-wallet-connect-next/blob/master/http_service.go#L44) and on the [frontend](https://github.com/getAlby/nostr-wallet-connect-next/blob/master/frontend/platform_specific/http/src/utils/request.ts#L7) 

Before the app starts up fully, the frontend and backend would both need to generate private-public keypairs and exchange their public keys, so that they can encrypt using the public key of the other end, which can use its private key to decrypt.

We need a secure algorithm that can easily be implemented on both the JS frontend and Go backend, and also implement the key exchange.

**Comment by bumi** [comment]
> Before the app starts up fully

I think this should be done on a session level (similar to a CSRF token). When the app is visited the key is generated and exchanged.

And I am wondering what standards/best-practices can be used here. 

On the backend this can be done with a echo middleware that decrypts the request body and encrypts the response body for API calls.
On the frontend this can be done with axios interceptors I think: https://axios-http.com/docs/interceptors - which decrypt the response body from the backend and encrypt the request body)

**Comment by rdmitr** [comment]
IMHO the first and foremost encryption best practice is you don't do it yourself 😁 The standard is TLS — can we leverage that?

I am not quite sure I understand the use case here, though. Is it intended for users who run NWC backend on a dedicated server/device and then connect to it from a local machine?

If yes, that sounds like a "power user" scenario to me. Would it be reasonable to expect such a user to be able to generate a self-signed TLS certificate by following fairly simple instructions? If yes again, I believe we may want to go TLS.

Also, echo supports "auto TLS" mode where it requests a letsencrypt certificate automatically. However, that would require the users to use a custom domain for their servers.

**Comment by rolznz** [comment]
> I am not quite sure I understand the use case here, though. Is it intended for users who run NWC backend on a dedicated server/device and then connect to it from a local machine?

That's right. But it's not a power user scenario, it would be for all Alby users. They generate or enter a seed phrase on the client and it must be sent to the server.

I think we need something different - to create a secure communication channel for a single session (not Alby users manually having to set something up). This would be a second, internal layer of encryption/decryption that encrypts payloads before they are sent, and decrypts them after they are received. We still have standard TLS for the requests.

This way, only the client and NWC backend can decrypt the payloads in a secure communication channel. It won't be enough to have the TLS certificate.

**Comment by rdmitr** [comment]
I'm still a bit confused 🤔 With TLS — unless it's compromised — it's the same: no one in the middle can decrypt the payload. Could you please elaborate a bit on why you feel TLS might be insufficient?

**Comment by rolznz** [comment]
You have a good point. I think Cloudflare can do TLS decryption (It is not enabled, but what if it was? how would anyone from the outside know by looking at the app source code?).

Users have to have some level of trust to use Alby-hosted apps anyway for multiple reasons. If they are uncomfortable with this they can self-host the application.

Thanks for looking into this. If it's sufficient as it is now, maybe we do not need to do anything. CC @bumi

**Comment by rolznz** [comment]
@rdmitr thanks for pushing back on this. For now We can rely on TLS for keeping the communication secure between the server and client but we will keep the issue open in case we have further requirements.
